pkg/controller/persistentvolumelabel/pvlcontroller.go

```diff
@@ -0,0 +1,187 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cloud
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/golang/glog"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/kubernetes/pkg/api"
+	coreinformers "k8s.io/kubernetes/pkg/client/informers/informers_generated/externalversions/core/v1"
+	"k8s.io/kubernetes/pkg/cloudprovider"
+	"k8s.io/kubernetes/pkg/cloudprovider/providers/aws"
+	"k8s.io/kubernetes/pkg/cloudprovider/providers/gce"
+	vol "k8s.io/kubernetes/pkg/volume"
+)
+
+type PersistentVolumeLabelController struct {
+	pvlInformer      coreinformers.PersistentVolumeInformer
+	ebsVolumes       aws.Volumes
+	gceCloudProvider *gce.GCECloud
+	cloud            cloudprovider.Interface
+	mutex            sync.Mutex
+}
+
+const ()
+
+// NewPersistentVolumeLabelController creates a PersistentVolumeLabelController object
+func NewPersistentVolumeLabelController(
+	pvlInformer coreinformers.PersistentVolumeInformer,
+	cloud cloudprovider.Interface) *PersistentVolumeLabelController {
+	pvlc := &PersistentVolumeLabelController{
+		pvlInformer: pvlInformer,
+		cloud:       cloud,
+	}
+
+	pvlInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: pvlc.AddLabels,
+	})
+	return pvlc
+}
+
+// This controller adds labels to persistent volumes and sets them to
+// an available state
+func (pvl *PersistentVolumeLabelController) Run(stopCh <-chan struct{}) {
+	go func() {
+		defer utilruntime.HandleCrash()
+		<-stopCh
+	}()
+}
+
+// AddLabels adds appropriate labels to persistent volumes and sets the
+// volume as available if successful.
+func (pvlc *PersistentVolumeLabelController) AddLabels(obj interface{}) {
+	volume, ok := obj.(*api.PersistentVolume)
+	if !ok {
+		return
+	}
+	var volumeLabels map[string]string
+	if volume.Spec.AWSElasticBlockStore != nil {
```

Comment:
They are currently mutually exclusive, but isn't it a desire to support multiple cloud providers in the future?